crypto/ec/ecp_secp384r1.c

```diff
@@ -0,0 +1,19594 @@
+                                         fiat_secp384r1_uint1 arg1,
+                                         uint64_t arg2, uint64_t arg3) {
+    fiat_secp384r1_uint128 x1;
+    uint64_t x2;
+    fiat_secp384r1_uint1 x3;
+    x1 = ((arg1 + (fiat_secp384r1_uint128)arg2) + arg3);
+    x2 = (uint64_t)(x1 & UINT64_C(0xffffffffffffffff));
+    x3 = (fiat_secp384r1_uint1)(x1 >> 64);
+    *out1 = x2;
+    *out2 = x3;
+}
+
+/*
+ * The function fiat_secp384r1_subborrowx_u64 is a subtraction with borrow.
+ * Postconditions:
+ *   out1 = (-arg1 + arg2 + -arg3) mod 2^64
+ *   out2 = -⌊(-arg1 + arg2 + -arg3) / 2^64⌋
+ *
+ * Input Bounds:
+ *   arg1: [0x0 ~> 0x1]
+ *   arg2: [0x0 ~> 0xffffffffffffffff]
+ *   arg3: [0x0 ~> 0xffffffffffffffff]
+ * Output Bounds:
+ *   out1: [0x0 ~> 0xffffffffffffffff]
+ *   out2: [0x0 ~> 0x1]
+ */
+static void fiat_secp384r1_subborrowx_u64(uint64_t *out1,
+                                          fiat_secp384r1_uint1 *out2,
+                                          fiat_secp384r1_uint1 arg1,
+                                          uint64_t arg2, uint64_t arg3) {
+    fiat_secp384r1_int128 x1;
+    fiat_secp384r1_int1 x2;
+    uint64_t x3;
+    x1 = ((arg2 - (fiat_secp384r1_int128)arg1) - arg3);
+    x2 = (fiat_secp384r1_int1)(x1 >> 64);
+    x3 = (uint64_t)(x1 & UINT64_C(0xffffffffffffffff));
+    *out1 = x3;
+    *out2 = (fiat_secp384r1_uint1)(0x0 - x2);
+}
+
+/*
+ * The function fiat_secp384r1_mulx_u64 is a multiplication, returning the full double-width result.
+ * Postconditions:
+ *   out1 = (arg1 * arg2) mod 2^64
+ *   out2 = ⌊arg1 * arg2 / 2^64⌋
+ *
+ * Input Bounds:
+ *   arg1: [0x0 ~> 0xffffffffffffffff]
+ *   arg2: [0x0 ~> 0xffffffffffffffff]
+ * Output Bounds:
+ *   out1: [0x0 ~> 0xffffffffffffffff]
+ *   out2: [0x0 ~> 0xffffffffffffffff]
+ */
+static void fiat_secp384r1_mulx_u64(uint64_t *out1, uint64_t *out2,
+                                    uint64_t arg1, uint64_t arg2) {
+    fiat_secp384r1_uint128 x1;
+    uint64_t x2;
+    uint64_t x3;
+    x1 = ((fiat_secp384r1_uint128)arg1 * arg2);
+    x2 = (uint64_t)(x1 & UINT64_C(0xffffffffffffffff));
+    x3 = (uint64_t)(x1 >> 64);
+    *out1 = x2;
+    *out2 = x3;
+}
+
+/*
+ * The function fiat_secp384r1_cmovznz_u64 is a single-word conditional move.
+ * Postconditions:
+ *   out1 = (if arg1 = 0 then arg2 else arg3)
+ *
+ * Input Bounds:
+ *   arg1: [0x0 ~> 0x1]
+ *   arg2: [0x0 ~> 0xffffffffffffffff]
+ *   arg3: [0x0 ~> 0xffffffffffffffff]
+ * Output Bounds:
+ *   out1: [0x0 ~> 0xffffffffffffffff]
+ */
+static void fiat_secp384r1_cmovznz_u64(uint64_t *out1,
+                                       fiat_secp384r1_uint1 arg1, uint64_t arg2,
+                                       uint64_t arg3) {
+    fiat_secp384r1_uint1 x1;
+    uint64_t x2;
+    uint64_t x3;
+    x1 = (!(!arg1));
+    x2 = ((fiat_secp384r1_int1)(0x0 - x1) & UINT64_C(0xffffffffffffffff));
+    x3 = ((x2 & arg3) | ((~x2) & arg2));
+    *out1 = x3;
+}
```

Comment:
This function basically does the same as constant_time_select_64(), except you pass it 1/0 instead of the mask (x2). We had to change our code to add a value_barrier_64() call around the mask/x2, to prevent gcc from recognizing it as a select and turning it into a branch.

Reply:
> prevent gcc from recognizing it as a select and turning it into a branch

Yea there's a [fiat-crypto issue](https://github.com/mit-plv/fiat-crypto/issues/748) for this. I have to think about the best way to handle it in ECCKiila. (And IIUC, it's `clang` not `gcc`.)

Reply:
I'm pretty sure we saw the effect with gcc, maybe clang does it too.

Reply:
I've merged https://github.com/mit-plv/fiat-crypto/pull/856, which autogenerates a value barrier here.

Reply:
TYVM @JasonGross! I'll test it out now

Reply:
So at least the ECCKiila testing for P-384 went fine, as well as the OpenSSL testing with this PR rebased in d63e8fe0f1c6687f3432064d8d52399d7f75a434. Let's see what the OpenSSL CI says in a few hours.

Reply:
CI went swell so I'm resolving this. Thanks @JasonGross for the fix fiat-side and @kroeckx for raising the issue.